File: js_example/src/did_utils.js
import bs58 from 'bs58';
import { blake2b } from '@noble/hashes/blake2b';

export function deriveDidId(genesisHashHex, publicKey) {
  const genesisBytes = hexToBytes(stripHex(genesisHashHex));
  const material = concatBytes(
    new TextEncoder().encode('QSB_DID'),
    genesisBytes,
    publicKey
  );
  const didIdBytes = blake2b(material, { dkLen: 32 });
  return bs58.encode(didIdBytes);
}

export function deriveSchemaId(genesisHashHex, schemaJson) {
  const genesisBytes = hexToBytes(stripHex(genesisHashHex));
  const material = concatBytes(
    new TextEncoder().encode('QSB_SCHEMA'),
    genesisBytes,
    schemaJson
  );
  const schemaIdBytes = blake2b(material, { dkLen: 32 });
  const schemaIdB58 = bs58.encode(schemaIdBytes);
  return `did:qsb:schema:${schemaIdB58}`;
}

export function stripHex(hex) {
  return hex.startsWith('0x') ? hex.slice(2) : hex;
}

export function hexToBytes(hex) {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i += 1) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

export function concatBytes(...parts) {
  const total = parts.reduce((sum, p) => sum + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

File: python_example/src/app/did_utils.py
import base58
import hashlib


def derive_did_id(genesis_hash_hex: str, public_key: bytes) -> str:
    genesis_bytes = bytes.fromhex(genesis_hash_hex.removeprefix("0x"))
    material = b"QSB_DID" + genesis_bytes + public_key
    did_id_bytes = hashlib.blake2b(material, digest_size=32).digest()
    return base58.b58encode(did_id_bytes).decode("ascii")


def derive_schema_id(genesis_hash_hex: str, schema_json: bytes) -> str:
    genesis_bytes = bytes.fromhex(genesis_hash_hex.removeprefix("0x"))
    material = b"QSB_SCHEMA" + genesis_bytes + schema_json
    schema_id_bytes = hashlib.blake2b(material, digest_size=32).digest()
    schema_id_b58 = base58.b58encode(schema_id_bytes).decode("ascii")
    return f"did:qsb:schema:{schema_id_b58}"
